### ISSUE TITLE
NBTItem problem

### ISSUE THREAD

**Issue by Kyumar** [tr7zw/Item-NBT-API#161]
Server version: 1.12.2
api version: 2.8.0

I am applying a nbt to an item using NBTItem when the item is dropped, everything works correctly, dropping an item with another nbt (applied by another plugin) dropping the item you lose everything, name, lore etc, ideas?

**Comment by tr7zw** [comment]
Check that you are updating the item in the inventory. NBTItem is a copy of the item in the inventory. Also print out the data before and after dropping. Noteworthy that the nbtapi has no event hooks, so basically has to be another plugin screwing up.

**Comment by Kyumar** [comment]
when player drop item, execute that

    public void apply(ItemStack itemStack, int id){
        NBTItem nbtItem = new NBTItem(itemStack);

        nbtItem.setInteger("dropID", id);
        nbtItem.applyNBT(itemStack);
    }

when i pickup it nothing changes, when i test it in another server, if i drop an item with an nbt, all data disappears

**Comment by Kyumar** [comment]
without adding back those who were there?

**Comment by tr7zw** [comment]
It sets the entire nbt of the itemstack to that of NTBItem.

**Comment by Kyumar** [comment]
Ok, thank you so much

**Comment by Kyumar** [comment]
So, if i have to add an nbt to an item who might have custom nbt should i use mergeCustomNBT# ?

**Comment by tr7zw** [comment]
You best case don't use any of these methods. Either use .getItem() to get back the modified item, or add ``true`` to the constructor of NBTItem to enable direct apply mode.

**Comment by Kyumar** [comment]
public ItemStack apply(ItemStack itemStack, int id){
        NBTItem nbtItem = new NBTItem(itemStack, true);

        nbtItem.setInteger("dropID", id);
        nbtItem.mergeCustomNBT(itemStack);
        return nbtItem.getItem();
    }

something like that?

**Comment by tr7zw** [comment]
No. 
```Java
public void apply(ItemStack itemStack, int id){
NBTItem nbtItem = new NBTItem(itemStack, true);
    nbtItem.setInteger("dropID", id);
}
```

**Comment by Kyumar** [comment]
oh, ok

**Comment by Kyumar** [comment]
Thank you so much

**Comment by Kyumar** [comment]
Hey sorry, even doing so continues to remove the other nbt tags and put only dropID

**Comment by Kyumar** [comment]
https://www.toptal.com/developers/hastebin/onaramudux.java that's the apply# code

https://www.toptal.com/developers/hastebin/ebaqulipol.csharp that's the drop event code where is fired apply#

**Comment by tr7zw** [comment]
Ehhhh that code shouldn't work at all? You are trying to update the event after it happened. Also, you could just store all of the data you are putting into the database onto the item^^

**Comment by Kyumar** [comment]
i have to get dropID when a player pickup an item and get the ID

**Comment by tr7zw** [comment]
Yes, but your event handler already technically shouldn't work and can have really bad side effects, because you are using an (async) task there.

**Comment by Kyumar** [comment]
so i shouldn't apply a nbt to the drop to account for it?

**Comment by tr7zw** [comment]
Not retroactively.

**Comment by Kyumar** [comment]
Quindi, la soluzione quale potrebbe essere? Al drop genero un id ed inserisco tutto nel db, per tenere conto che a quel drop è associato quel determinato id come faccio?

**Comment by Kyumar** [comment]
So what could the solution be? At the drop I generate an id and insert everything in the db, to take into account that that particular id is associated with that drop, how do I do it?

**Comment by tr7zw** [comment]
I mean what are you even trying to achieve? Because at pickup you have the reverse issue of your SQL query will happen in the future after the person/hopper/entity has picked up the item.

**Comment by tr7zw** [comment]
Again, why do it this way then xD On drop write to the entity!(not the item stack) who dropped it and when. And at pickup get that data + who picked it up and when and write it async to the database.

**Comment by tr7zw** [comment]
Ah wait you are in 1.12, then you need to write it to the item instead of the entity.

**Comment by tr7zw** [comment]
The issue is that your code breaks the spigot API logic. Please add a few
``System.out.println(new NBTItem(event.getItemStack));`` to a) before you do anything b) your async logic c) after updating the item.

**Comment by Kyumar** [comment]
done, apply and correctly take the right nbt

**Comment by Kyumar** [comment]
Ok, i think it's not a problem of NBTApi ...
